pandas/core/indexes/interval.py

```diff
@@ -1090,6 +1090,37 @@ def equals(self, other):
     def overlaps(self, other):
         return self._data.overlaps(other)
 
+    def intersection(self, other, sort=False):
+        other = self._as_like_interval_index(other)
+
+        # GH 19016: ensure set op will not return a prohibited dtype
+        subtypes = [self.dtype.subtype, other.dtype.subtype]
+        common_subtype = find_common_type(subtypes)
+        if is_object_dtype(common_subtype):
+            msg = ('can only do intersection between two IntervalIndex '
+                   'objects that have compatible dtypes')
+            raise TypeError(msg)
+
+        if self.left.is_unique and self.right.is_unique:
+            lindexer = self.left.get_indexer(other.left)
+            rindexer = self.right.get_indexer(other.right)
+            match = (lindexer == rindexer) & (lindexer != -1)
+            indexer = lindexer.take(match.nonzero()[0])
+            taken = self.take(indexer)
+        else:
+            # duplicates
+            lmiss = other.left.get_indexer_non_unique(self.left)[1]
```

Comment:
this is really complex can u split to a searate function (each case)

Reply:
`_​intersection_unique` and `_intersection_non_unique` works in an opposite way.

`_intersecton_unique` gathers the matches on both hand sides. `_intersection_non_unique` eliminates the unmatched on the left hand side first and then the unmatched on the right hand side.